### DISCUSSION TITLE
PowerShell SDK support for MAUI

### DISCUSSION THREAD

**Discussion by rhubarb-geek-nz** [PowerShell/PowerShell#20840] [Enhancements]
Can the PowerShell SDK be used with [MAUI](https://learn.microsoft.com/en-us/dotnet/maui/what-is-maui?view=net-maui-8.0)?

[#17941 ](https://github.com/PowerShell/PowerShell/issues/17941#issuecomment-1837541173)

The discussion was around supporting Android. Given MAUI is already doing Android then fitting in with this runtime would be better than having multiple PowerShell solutions for Android.

There are a few issues, from the documents it looks like MAUI is based on Mono Runtime rather than .NET CoreCLR.

**Comment by fMichaleczek** [comment]
I've successfully compiled PowerShell 7.0 to WebAssembly on the Mono runtime and tested an application on Android back then. However, the challenge extends beyond mere compilation. The core issue lies in the absence of a robust application model for PowerShell. While traces exist in PSISE, the non-open-source code, which includes fragments of Visual Studio, complicates adaptation to modern UI standards in 2023.

One potential approach involves revisiting concepts from PSISE, such as those utilized with $PSISE, and aligning them with contemporary UI practices. A promising solution could be leveraging Avalonia UI, which offers clarity compared to the array of UI technologies Microsoft has introduced (WinForms, WPF, UWP, etc.).

Microsoft appears entangled in a myriad of solutions, including attempts to integrate different UI hosting generations, leading to a convoluted mix of technologies. Personally, I've disengaged from keeping up with these developments. On the contrary, Avalonia is relatively new but based on WPF, benefiting from community feedback on application modeling.

I'm anticipating that DotNet 9.0 will address multithreading challenges in WebAssembly, potentially paving the way for a cross-platform Avalonia framework supporting Windows/Linux/MacOS/Android/Browser. However, it's worth noting that accessing modules on Android and Browser platforms without requiring recompilation poses a significant hurdle, as PowerShell currently lacks application-oriented modules for these platforms, necessitating extensive development efforts.

**Reply by fMichaleczek** [reply]
This Avalonia Terminal Emulator UI is MIT License :
https://github.com/VitalElement/AvalonStudio.TerminalEmulator

Avalonia Studio is GPLv3 but we are able to imagine what could be a PowerShell Avalonia ISE 
![image](https://github.com/PowerShell/PowerShell/assets/6187529/ec79737c-7e7d-4abc-abef-dfdf5870f9d4)

The first app in the model we need develop is PowerShell Avalonia ISE to build PowerShell Avalonia apps

**Comment by rhubarb-geek-nz** [comment]
When I think of the PowerShell SDK I am not thinking in terms of UI at all.

What I am looking for is can I create an InitialSessionState, call PowerShell.Create(initialState) and then call Invoke on the powershell object itself.

So can I use the SDK in a MAUI application so that I can implement parts of the application logic using PowerShell,

I would not dream of producing an Android app where I was expecting the user to write PowerShell.

**Reply by fMichaleczek** [reply]
https://twitter.com/_Flavien/status/1281715079831064576/photo/1

**Comment by rhubarb-geek-nz** [comment]
To clarify, I see PowerShell as a powerful scripting engine and I can see plenty of benefits writing pipelines to perform sequenced actions and workflow within the app, but that is just internal implementation, the mobile user does not need to know or care. 

Just like the user of a mobile app does not care if it is written in Java, Kotlin, React-Native, C#, Swift or objective-C. To them it is just an app.

In exactly the same way, I could write [web-services in PowerShell](https://github.com/rhubarb-geek-nz/AspNetForPowerShell/blob/main/DemoApp/DemoApp.ps1), but the person using a web-browser would have no idea.

**Reply by fMichaleczek** [reply]
The issue is that your mindset is too Win32-oriented to grasp the challenge. Let's be factual to illustrate it:

- There is no System.Console in the Android and Browser Application Model.
- The absence of a console means that creating a host requires a significant amount of work.
- No console exists because there is no accessible tty in the OS application model.
- The threading model for modern apps is a single thread, and PowerShell requires 2 threads to function correctly (1 for the interpreter and 1 for handling asynchronous tasks), consequently.
- Linux is not implemented in MAUI.

What can be done:
- Implement an app without model (using the framework of your choice) and assume that it will start SMA and execute a script at launch.
- Implement a backend system as if PowerShell handle a controller.
- Implement basic modules for a UI (Drawing, Model, View, etc.).

In any case, MAUI is the worst candidate for PowerShell. 
Xaml Island is terrible (it's multi-host Xaml with routing between different engines, each with its own set of problems!). 

Avalonia has everything that is lacking, and if you are motivated, you can look at the application model of PSISE.

**Reply by rhubarb-geek-nz** [reply]
> The issue is that your mindset is too Win32-oriented to grasp the challenge. Let's be factual to illustrate it:

I beg your pardon?

- No console required - I envisaged the PowerShell cmdlets working completely headless
- Creating a host with no console still means there is still no console, it is headless, stdin is from /dev/null, stdout and stderr are to /dev/null
- no tty required
- there is no console interpreter required, however cmdlets will need to run in their individual threads in order to correctly pipeline

Now my understanding that MAUI on Linux uses the Mono implementation of .NET, and that happily provides threads.

So I would see all the UI still being done by MAUI+Xaml but just like it can invoke asynchronous network requests, it should also be able to initiate asynchronous PowerShell pipelines. The PowerShell pipelines do not interact with the UI at all, they do what they always do, which is read from the input pipe and write to the output pipe.

Now the show-stopper might be that it is too much work to use Mono rather than .NET Core. But given PowerShell 5 runs on .NET framework I could imagine being able to port the core engine to Mono with the goal of keeping the binary API the same.

**Reply by fMichaleczek** [reply]
Only if you approach it with the Product/Platform mindset because that's what leads to the KO. It's not just a technical challenge; there's Microsoft being troublesome with its SDK vision of SDK.

MAUI only recognizes operating systems that have an Application SDK (thus an application model). Since Linux is a kernel and the GTK 4.0 Application SDK doesn't exist for C#, there's nothing.

What you're envisioning is still a Windows/Linux-centric view. You can compile an APK under Android, but it will run with the Android SDK, thus in a sandbox. And in a sandbox, you can't do whatever you want.

For task management, you can use this API:
https://learn.microsoft.com/en-us/dotnet/standard/asynchronous-programming-patterns/interop-with-other-asynchronous-patterns-and-types?redirectedfrom=MSDN#from-apm-to-tap

Otherwise, check out my PR if you want to see what patch I needed for WASM:
https://github.com/fMichaleczek/EmbeddedPowerShell/blob/29b4ca929e5eb2d729aa251baa12833a87eeef19/demos/Blazor/powershell-blazor-wasm-demo/Pages/PowerShell.razor"

you don't react to my link to Twitter in the other response, so I show you the proof it's possible.
![image](https://github.com/PowerShell/PowerShell/assets/6187529/60b6db9a-28f5-4d04-8e9f-50c7d3eb085c)

**Comment by rhubarb-geek-nz** [comment]
> What you're envisioning is still a Windows/Linux-centric view. You can compile an APK under Android, but it will run with the Android SDK, thus in a sandbox. And in a sandbox, you can't do whatever you want.

Yes, I am deliberately thinking about the Sandboxed world on Android and iPhone. The end user expects apps to be self-contained and restricted by those operating systems and that they remain in control over the app's permissions.

**Reply by fMichaleczek** [reply]
- Windows : OK but will only work on Windows 10 supported branch and not on Windows Server
- Linux : No, there is no MAUI
- MacOS : should work
- Android : yes but in a sandbox
- iphone :  should not work with PowerShell requirement - dotnet requires precompiled app (using Mono AOT) on iphone  because Apple Policy
- Browser : No (but you can host a browser)
